### PR TITLE
fix etcd full backup will always alert when create new shoot

### PR DESCRIPTION
**Pull request by zjj2wry**
**How to categorize this PR?**

/area monitoring
/kind enhancement
/priority normal

**What this PR does / why we need it**:

the `KubeEtcdFullBackupFailed` will always alert when creating a new shoot. I hope this just alerts when etcdbr_snapshot_latest_revision!=0

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @zjj2wry for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by zjj2wry** [comment]
Fixes https://github.com/gardener/gardener/issues/2894

**Comment by rfranzke** [comment]
Thanks for the PR!
/assign @wyb1 @istvanballok 
/reviewed ok-to-test

**Review by wyb1 — COMMENTED** [review]
Thanks for the PR @zjj2wry!
Please adjust the alert unit test found [here](https://github.com/gardener/gardener/blob/4af3eb3457649ec37f67a29695c62708bbe490f9/charts/seed-monitoring/charts/core/charts/prometheus/rules-tests/kube-etcd3.rules.test.yaml#L139)

**Comment by shreyas-s-rao** [comment]
@zjj2wry Thanks for your PR! Ideally, this alert shouldn't be firing at shoot creation, since the first thing etcd-backup-restore does is take a full snapshot of the freshly bootstrapped etcd. If the alert is firing, that probably means that backup sidecar isn't able to take/upload a snapshot of the etcd to the configured object storage bucket (probably due to misconfigured credentials or network issues), which would mean the etcd would be at risk of data loss on possible data corruption.

**Comment by zjj2wry** [comment]
`time() - etcdbr_snapshot_latest_timestamp{job="kube-etcd3-backup-restore",kind="Full",role="main"} > bool 86400` always true when etcd startup, this will alert when etcd restart

**Comment by shreyas-s-rao** [comment]
@zjj2wry Partially yes, that expression will evaluate to true, but only on fresh etcd start or if full snapshot hasn't been taken for more than a day. We also combine this check with `etcdbr_snapshot_required{kind="Full", role="main"}` so that the alert is triggered only if a full snapshot is actually required (which evaluates to true when etcd and backup sidecar start, but once full snapshot is uploaded, evaluates to false). So the alert would fire only if full snapshot wasn't taken for 15 minutes, which is the alert evaluation period. Can you please check whether backup sidecar is able to upload snapshots to the configured bucket? Or if you have disabled backups, check the health of the etcd pod. You can see the logs of the `backup-restore` container for more information regarding the backups being or not being taken.

**Comment by rfranzke** [comment]
@zjj2wry can you please comment on the concerns of @shreyas-s-rao?

**Comment by zjj2wry** [comment]
> So the alert would fire only if full snapshot wasn't taken for 15 minutes

We currently do a full backup every 24 hours. If etcd is restarted, will it be backed up immediately? or wait 24h?

**Comment by zjj2wry** [comment]
>  for now due to inactivity, feel free to reopen if you think it is necessary

can you help reopen this, seems I do not have permissions. I am on vacation last week because of China National Day

**Comment by zjj2wry** [comment]
@shreyas-s-rao this alert will trigger when ETCD restart, would you like to verify it?

**Comment by amshuman-kr** [comment]
> We currently do a full backup every 24 hours. If etcd is restarted, will it be backed up immediately? or wait 24h?

I think it will wait till `24h` for the full backup. https://github.com/gardener/etcd-backup-restore/blob/master/pkg/server/backuprestoreserver.go#L187-L231

**Comment by zjj2wry** [comment]
> > We currently do a full backup every 24 hours. If etcd is restarted, will it be backed up immediately? or wait 24h?
> 
> I think it will wait till `24h` for the full backup. https://github.com/gardener/etcd-backup-restore/blob/master/pkg/server/backuprestoreserver.go#L187-L231

tks, I will receive an alert when ETCD restarts in my env.

**Comment by shreyas-s-rao** [comment]
@zjj2wry I'm guessing the etcd restart takes more than 15 minutes, due to which prometheus isn't able to scrape `etcdbr_` metrics for more than 15 minutes, which triggers the alert. The alert wouldn't fire immediately on etcd restart.

**Comment by rfranzke** [comment]
@zjj2wry can you confirm this?

**Comment by zjj2wry** [comment]
> @zjj2wry I'm guessing the etcd restart takes more than 15 minutes, due to which prometheus isn't able to scrape `etcdbr_` metrics for more than 15 minutes, which triggers the alert. The alert wouldn't fire immediately on etcd restart.

why 15 minutes? Where is configured for 15 minutes. and the backup job will start after 24h? so will always alert in 15min~24h?

**Comment by shreyas-s-rao** [comment]
@zjj2wry FYI
> Where is configured for 15 minutes

The `for` field [here](https://github.com/gardener/gardener/blob/2cca55296a4b0e92b90325ca7dfdb79f9ba6e444/charts/seed-monitoring/charts/core/charts/prometheus/rules/kube-etcd3.rules.yaml#L107) specifies how long the alert condition has to be true before actually firing an alert for the rule.

> the backup job will start after 24h?

No. The first backup will be taken conditionally as soon as the sidecar starts, based on the following policy: "if there has NOT been a full snapshot in the last 24 hours, take a full snapshot. Else, schedule the next full snapshot according to the given cron schedule (passed to the `etcdbrctl` command via the command line parameter `--schedule`)". The alert is also configured to use the `etcdbr_snapshot_required` metric, which specifies whether a snapshot is *actually* required at that moment or not.

For a fresh etcd, as soon as the etcd pod is up, an initial full snapshot is taken, and the next full snapshot is configured according to the cron schedule that has been set. As soon as the full snapshot is taken, `etcdbr_snapshot_required` is set to false, until etcd receives some writes/updates (PUT or DELETE), after which `etcdbr_snapshot_required` is set to true.

In other words, this alert will only fire if there has been no full snapshot for the past 24 hours **and** etcd has received some new data which has not been backed up yet. So if the alert is firing for you on etcd startup, that means the backup sidecar isn't able to upload a snapshot to the specified backup storage bucket. Can you please confirm this? You may refer to the `etcd-main-0` pod's `backup-restore` container logs, as well as the specified storage bucket for more insight into the backups.

**Comment by zjj2wry** [comment]
@shreyas-s-rao 

delete etcd pod, search the expr will true
```
(time() - etcdbr_snapshot_latest_timestamp{job="kube-etcd3-backup-restore",kind="Full",role="main"} > bool 86400) + (etcdbr_snapshot_required{kind="Full", role="main"} >= bool 1) == 2
```
image

the timestamp of full backup is 0
```
etcdbr_snapshot_latest_timestamp{job="kube-etcd3-backup-restore",kind="Full",role="main",cluster="shoot/gcp/us-west1/shoot--staging--3c107348"} 
```
image

this is our backup-restore log, seems not doing full backup
```bash
➜  ~ k logs -f etcd-main-0 -n shoot--staging--3c107348 -c backup-restore
time="2020-10-19T10:06:31Z" level=info msg="etcd-backup-restore Version: v0.9.0"
time="2020-10-19T10:06:31Z" level=info msg="Git SHA: 105deb66"
time="2020-10-19T10:06:31Z" level=info msg="Go Version: go1.12.7"
time="2020-10-19T10:06:31Z" level=info msg="Go OS/Arch: linux/amd64"
time="2020-10-19T10:06:31Z" level=info msg="defragmentationSchedule: 14 19 */3 * *\netcdConnectionConfig:\n  caFile: /var/etcd/ssl/ca/ca.crt\n  certFile: /var/etcd/ssl/client/tls.crt\n  connectionTimeout: 5m0s\n  endpoints:\n  - https://etcd-main-local:2379\n  keyFile: /var/etcd/ssl/client/tls.key\nrestorationConfig:\n  embeddedEtcdQuotaBytes: 8589934592\n  initialAdvertisePeerURLs:\n  - http://localhost:2380\n  initialCluster: default=http://localhost:2380\n  initialClusterToken: etcd-cluster\n  maxFetchers: 6\n  name: default\n  restoreDataDir: /var/etcd/data/new.etcd\nserverConfig:\n  port: 8080\n  server-cert: /var/etcd/ssl/server/tls.crt\n  server-key: /var/etcd/ssl/server/tls.key\nsnapshotterConfig:\n  deltaSnapshotMemoryLimit: 104857600\n  deltaSnapshotPeriod: 5m0s\n  garbageCollectionPeriod: 12h0m0s\n  garbageCollectionPolicy: LimitBased\n  maxBackups: 7\n  schedule: 59 17 * * *\nsnapstoreConfig:\n  container: \"\"\n  maxParallelChunkUploads: 5\n  prefix: shoot--staging--3c107348--849357e8-4900-4c1d-9f37-611e1426e9ac/etcd-main/v1\n  provider: GCS\n  tempDir: /var/etcd/data/temp\n"
time="2020-10-19T10:06:31Z" level=info msg="Creating snapstore from provider: GCS" actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Creating snapshotter..." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Setting status to : 503" actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Registering the http request handlers..." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Starting the http server..." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Probing etcd..." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Starting HTTP server at addr: :8080" actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="TLS enabled. Starting HTTPS server." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Responding to status request with: New" actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Received start initialization request." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Updating status from New to Progress" actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Changed handler state." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Waiting for acknowledgment..." actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Validation failBelowRevision: " actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Validation mode: sanity" actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Checking for data directory structure validity..."
time="2020-10-19T10:06:31Z" level=info msg="Responding to status request with: Progress" actor=backup-restore-server
time="2020-10-19T10:06:31Z" level=info msg="Checking for revision consistency..."
time="2020-10-19T10:06:32Z" level=info msg="Data directory valid."
time="2020-10-19T10:06:32Z" level=info msg="Successfully initialized data directory for etcd." actor=backup-restore-server
time="2020-10-19T10:06:32Z" level=info msg="Responding to status request with: Successful" actor=backup-restore-server
time="2020-10-19T10:06:32Z" level=info msg="Updating status from Successful to New" actor=backup-restore-server
time="2020-10-19T10:06:34Z" level=info msg="Applied watch on etcd from revision: 1517112" actor=snapshotter
time="2020-10-19T10:06:34Z" level=info msg="Taking delta snapshot for time: 2020-10-19 10:06:34.845223308 +0000 UTC" actor=snapshotter
time="2020-10-19T10:06:34Z" level=info msg="Uploading snapshot of size: 1214787, chunkSize: 5242880, noOfChunks: 1"
time="2020-10-19T10:06:34Z" level=info msg="Triggered chunk upload for all chunks, total: 1"
time="2020-10-19T10:06:34Z" level=info msg="No of Chunks:= 1"
time="2020-10-19T10:06:34Z" level=info msg="Uploading chunk with offset : 0, attempt: 0"
time="2020-10-19T10:06:34Z" level=info msg="Received chunk result for id: 1, offset: 0"
time="2020-10-19T10:06:34Z" level=info msg="Received successful chunk result for all chunks. Stopping workers."
time="2020-10-19T10:06:34Z" level=info msg="All chunk uploaded successfully. Uploading composite object."
time="2020-10-19T10:06:34Z" level=info msg="Composite object uploaded successfully."
time="2020-10-19T10:06:34Z" level=info msg="Total time to save delta snapshot: 0.115302 seconds."
time="2020-10-19T10:06:34Z" level=info msg="Successfully saved delta snapshot at: Backup-1603043940/Incr-01517112-01518201-1603101994" actor=snapshotter
time="2020-10-19T10:06:34Z" level=info msg="Setting status to : 200" actor=backup-restore-server
time="2020-10-19T10:06:34Z" level=info msg="Starting snapshotter..." actor=backup-restore-server
time="2020-10-19T10:06:34Z" level=info msg="Will take next full snapshot at time: 2020-10-19 17:59:00 +0000 UTC" actor=snapshotter
time="2020-10-19T10:11:34Z" level=info msg="Taking delta snapshot for time: 2020-10-19 10:11:34.965304551 +0000 UTC" actor=snapshotter
time="2020-10-19T10:11:34Z" level=info msg="Uploading snapshot of size: 1283321, chunkSize: 5242880, noOfChunks: 1"
time="2020-10-19T10:11:34Z" level=info msg="Triggered chunk upload for all chunks, total: 1"
time="2020-10-19T10:11:34Z" level=info msg="No of Chunks:= 1"
time="2020-10-19T10:11:34Z" level=info msg="Uploading chunk with offset : 0, attempt: 0"
time="2020-10-19T10:11:35Z" level=info msg="Received chunk result for id: 1, offset: 0"
time="2020-10-19T10:11:35Z" level=info msg="Received successful chunk result for all chunks. Stopping workers."
time="2020-10-19T10:11:35Z" level=info msg="All chunk uploaded successfully. Uploading composite object."
time="2020-10-19T10:11:35Z" level=info msg="Composite object uploaded successfully."
time="2020-10-19T10:11:35Z" level=info msg="Total time to save delta snapshot: 0.173555 seconds."
time="2020-10-19T10:11:35Z" level=info msg="Successfully saved delta snapshot at: Backup-1603043940/Incr-01518202-01519428-1603102294" actor=snapshotter
time="2020-10-19T10:11:35Z" level=info msg="Stopping delta snapshot..." actor=snapshotter
time="2020-10-19T10:11:35Z" level=info msg="Resetting delta snapshot to run after 5m0s." actor=snapshotter
```

**Comment by shreyas-s-rao** [comment]
@zjj2wry I see you're using an older version of Gardener, which uses etcd-backup-restore v0.9.0, which had a bug in metric initialization logic. This has been fixed [here](https://github.com/gardener/etcd-backup-restore/pull/223), and is part of etcd-backup-restore v0.10.0. Please consider upgrading your Gardener setup to use this fix.

**Comment by zjj2wry** [comment]
> @zjj2wry I see you're using an older version of Gardener, which uses etcd-backup-restore v0.9.0, which had a bug in metric initialization logic. This has been fixed [here](https://github.com/gardener/etcd-backup-restore/pull/223), and is part of etcd-backup-restore v0.10.0. Please consider upgrading your Gardener setup to use this fix.

got it, I will close this issue.  thank you very much :))
